Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcCopyClient.cs

```diff
@@ -67,9 +129,27 @@ private GrpcCopyClient(string host, int grpcPort, bool useCompression)
         /// <param name="useCompression">Whether or not GZip is enabled for copies.</param>
         public static GrpcCopyClient Create(string host, int grpcPort, bool useCompression = false)
         {
-            // TODO: Add caching of GrpcCopyClient objects
-            // TODO: Add case where _clientPool has exceeded some maximum count
-            return new GrpcCopyClient(host, grpcPort, useCompression);
+            if (_clientDict.TryGetValue((host, grpcPort, useCompression), out GrpcCopyClient existingClient))
+            {
+                Interlocked.Increment(ref existingClient._uses);
+                existingClient._lastUseTime = DateTime.UtcNow;
+                return existingClient;
+            }
+            else if (_clientDict.Count > _maxClientCount)
+            {
+                throw new CacheException($"Attempting to create {nameof(GrpcCopyClient)} to increase cached count above maximum allowed ({_maxClientCount})");

```

Comment:
I think in this case we should clean-up first and only when all the clients are used we should fail.

Today do to an unbnormal spike in client creation you can get here and fail, even though you have a bunch of instances that are no longer used.

Reply:
I think this is fine for the time-being. If we end up with too many unused clients that last too long, then we can fix for the short-term via config options (reduce wait time between cleanups and reduce maximum age) and revisit this.

---
In reply to: [289209028](https://github.com/microsoft/BuildXL/pull/391#discussion_r289209028) [](ancestors = 289209028)